core/indexer/dataDispatcher.go

```diff
@@ -58,29 +61,40 @@ func (d *dataDispatcher) startWorker(ctx context.Context) {
 			log.Debug("dispatcher's go routine is stopping...")
 			return
 		case wi := <-d.chanWorkItems:
+			d.writeMutex.Lock()
 			d.doWork(wi, 0)
+			d.writeMutex.Unlock()
 		}
 	}
 }
 
 // Close will close the endless running go routine
 func (d *dataDispatcher) Close() error {
+	if d.wasClosed.IsSet() {
```

Comment:
concurrency problems here: check for set and the actual set should be done atomically. Worst case: use an uint32 value and functions found in `atomic` package